crates/chain/src/tx_graph.rs

```diff
@@ -681,11 +707,32 @@ impl<A: Anchor> TxGraph<A> {
             }
         };
 
+        // Since transactions that conflict with the root tx are also conflicts, we need to traverse
+        // all ancestor's conflicts. We only traverse unconfirmed ancestors since conflicts of
+        // confirmed transactions cannot be in the best chain.
+        let ancestor_txs =
+            TxAncestors::new_include_root(self, tx, |_, ancestor_tx: &Transaction| {
+                let tx_node = self.get_tx_node(ancestor_tx.txid())?;
+                for block in tx_node.anchors {
+                    match chain.is_block_in_chain(block.anchor_block(), chain_tip) {
+                        Ok(Some(true)) => return None,
+                        Err(e) => return Some(Err(e)),
+                        _ => continue,
+                    }
+                }
+                Some(Ok(tx_node.tx))
+            })
+            .collect::<Result<Vec<_>, C::Error>>()?;
+
+        let conflicting_txs = ancestor_txs
+            .into_iter()
+            .flat_map(|ancestor| self.walk_conflicts(ancestor, |_, txid| self.get_tx_node(txid)));
+
         // If a conflicting tx is in the best chain, or has `last_seen` higher than this tx, then
         // this tx cannot exist in the best chain
-        for conflicting_tx in self.walk_conflicts(tx, |_, txid| self.get_tx_node(txid)) {
-            for block in conflicting_tx.anchors.iter().map(A::anchor_block) {
-                if chain.is_block_in_chain(block, chain_tip)? == Some(true) {
+        for conflicting_tx in conflicting_txs {
```

Comment:
Agreed, I fixed by using `direct_conflicts_of_tx` instead of `walk_conflicts`, and considering the last_seen of each ancestor instead of the last_seen of the root tx:
```
-        let conflicting_txs = ancestor_txs
-            .into_iter()
-            .flat_map(|ancestor| self.walk_conflicts(ancestor, |_, txid| self.get_tx_node(txid)));
-
-        // If a conflicting tx is in the best chain, or has `last_seen` higher than this tx, then
-        // this tx cannot exist in the best chain
-        for conflicting_tx in conflicting_txs {
-            for block in conflicting_tx.anchors {
-                if chain.is_block_in_chain(block.anchor_block(), chain_tip)? == Some(true) {
-                    // conflicting tx is in best chain, so the current tx cannot be in best chain!
+        for tx_node in unconfirmed_ancestor_txs {
+            // We retrieve all the transactions conflicting with this specific ancestor
+            let conflicting_txs = self
+                .direct_conflicts_of_tx(tx_node.tx)
+                .filter_map(|(_, txid)| self.get_tx_node(txid));
+
+            // If a conflicting tx is in the best chain, or has `last_seen` higher than this ancestor, then
+            // this tx cannot exist in the best chain
+            // Note that `direct_conflicts_of_tx`, which we used above to retrieve conflicts, doesn't include conflicting
+            // descendants, so comparing `last_seen` is safe
+            for conflicting_tx in conflicting_txs {
+                for block in conflicting_tx.anchors {
+                    if chain.is_block_in_chain(block.anchor_block(), chain_tip)? == Some(true) {
+                        // conflicting tx is in best chain, so the current tx cannot be in best chain!
+                        return Ok(None);
+                    }
+                }
+                if conflicting_tx.last_seen_unconfirmed > tx_node.last_seen_unconfirmed {
                     return Ok(None);
                 }
             }
-            if conflicting_tx.last_seen_unconfirmed > *last_seen {
-                return Ok(None);
-            }

```

I added the scenario "B and B' spend A and conflict, C spends B, all the transactions are unconfirmed, B' has higher last_seen than B" in the conflicting tests, which would previously fail and now passes. Lmk if it seems correct to you!